pkg/postage/batchstore/store.go

```diff
@@ -5,62 +5,139 @@
+	unreserveFunc func(batchID []byte, radius uint8) error // unreserve function
 }
 
 // New constructs a new postage batch store.
-func New(st storage.StateStorer) postage.Storer {
-	return &store{st}
+// It initialises both chain state and reserve state from the persistent state store
+func New(st storage.StateStorer, unreserveFunc func(batchID []byte, radius uint8) error) (postage.Storer, error) {
+	cs := &postage.ChainState{}
+	err := st.Get(chainStateKey, cs)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return nil, err
+		}
+		cs = &postage.ChainState{
+			Block: 0,
+			Total: big.NewInt(0),
+			Price: big.NewInt(0),
+		}
+	}
+	rs := &reserveState{}
+	err = st.Get(chainStateKey, rs)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return nil, err
+		}
+		rs = &reserveState{
+			Depth:    DefaultDepth,
+			Core:     big.NewInt(0),
+			Edge:     big.NewInt(0),
+			Capacity: Capacity,
+		}
+	}
+	return &store{st, cs, rs, unreserveFunc}, nil
 }
 
 // Get returns a batch from the batchstore with the given ID.
 func (s *store) Get(id []byte) (*postage.Batch, error) {
 	b := &postage.Batch{}
 	err := s.store.Get(batchKey(id), b)
-
 	return b, err
 }
 
 // Put stores a given batch in the batchstore with the given ID.
-func (s *store) Put(b *postage.Batch) error {
+func (s *store) Put(b *postage.Batch, value *big.Int, depth uint8) error {
+	oldVal := big.NewInt(0).Set(b.Value)
+	oldDepth := b.Depth
+	err := s.store.Delete(valueKey(oldVal, b.ID))
+	if err != nil {
+		return err
+	}
+	b.Value.Set(value)
+	b.Depth = depth
+	err = s.store.Put(valueKey(b.Value, b.ID), nil)
+	if err != nil {
+		return err
+	}
+	err = s.updateValueChange(b, oldDepth, oldVal)
+	if err != nil {
+		return err
+	}
 	return s.store.Put(batchKey(b.ID), b)
 }
 
-// PutChainState implements BatchStorer. It stores the chain state in the batch
-// store.
-func (s *store) PutChainState(state *postage.ChainState) error {
-	return s.store.Put(stateKey, state)
+// delete removes the batches with ids given as arguments.
+func (s *store) delete(ids ...[]byte) error {
+	for _, id := range ids {
+		b, err := s.Get(id)
+		if err != nil {
+			return err
+		}
+		err = s.store.Delete(valueKey(b.Value, id))
+		if err != nil {
+			return err
+		}
+		err = s.store.Delete(batchKey(id))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-// GetChainState implements BatchStorer. It returns the stored chain state from
-// the batch store.
-func (s *store) GetChainState() (*postage.ChainState, error) {
-	cs := &postage.ChainState{}
-	err := s.store.Get(stateKey, cs)
-	if err == storage.ErrNotFound {
-		return &postage.ChainState{
-			Block: 0,
-			Total: big.NewInt(0),
-			Price: big.NewInt(0),
-		}, nil
+// PutChainState implements BatchStorer.
+// It purges expired batches and unreserves underfunded ones before it
+// stores the chain state in the batch store.
+func (s *store) PutChainState(cs *postage.ChainState) error {
+	s.cs = cs
```

Comment:
you're updating the state before the db change was written. it might return an error, and the component might be left in an inconsistent state

Reply:
i think that is correct. alternative is to quit the  node